### PR TITLE
Seperate integer/float divison - issue#109

### PR DESCRIPTION
**Pull request by ghost**
I've introduced an integer division operator,  <code>idiv</code> . I'm using this as <code>div</code> clashed with the libc 'div' external function - trivial to change. 

Would uppercase keyword operators be better? e.g. NOT, OR, AND, DIV, (MOD?)

### CODE REVIEW

**Comment by ghost** [comment]
Use <code>\</code> and <code>\=</code> for integer divide (quotient) operator.

File: compiler/src/lexer.cpp [hunk 1]
```diff
@@ -156,8 +156,8 @@ static bool keywordIdentifier(TokenPtr &x) {
     "<--", "-->", "..", "=>", "->",
     "==", "!=", "<=", ">=",
     "<", ">","++=","++",
-    "::", "+=", "-=", "*=", "/=", "%=",
-    "+", "-", "*", "/", "%", "=", "&", "^", "|",
+    "::", "+=", "-=", "*=", "/=", "\x5c\x3d", "%=",
```
Comment (jckarter) [inline compiler/src/lexer.cpp]: `"\\="` and `"\\"` would be a bit easier to read.

Comment (ghost) [inline compiler/src/lexer.cpp]: Didn't work with those.

File: compiler/src/parser.cpp [hunk 2]
```diff
@@ -650,6 +650,45 @@ static bool mulExpr(ExprPtr &x) {
     return true;
 }
 
+static bool quotientTail(VariadicOpPtr &x) {
+    LocationPtr location = currentLocation();
+    ExprListPtr exprs = new ExprList();
+    ExprPtr b;
+    int p = save();
+    if (!symbol("\x5c")) return false;
```
Comment (jckarter) [inline compiler/src/parser.cpp]: `"\\"`

File: lib-clay/cocoa/coreservices/generated/generated.macosx.x86.32.clay [hunk 3]
```diff
@@ -2050,9 +2050,9 @@ alias midiNoClientErr = 4294967046;
 alias midiNoPortErr = 4294967045;
 alias midiTooManyPortsErr = 4294967044;
 alias midiTooManyConsErr = 4294967043;
-alias midiVConnectErr = 4294967042;
-alias midiVConnectMade = 4294967041;
-alias midiVConnectRmvd = 4294967040;
+alias m\ConnectErr = 4294967042;
+alias m\ConnectMade = 4294967041;
+alias m\ConnectRmvd = 4294967040;
```
Comment (jckarter) [inline lib-clay/cocoa/coreservices/generated/generated.macosx.x86.32.clay]: oops

File: lib-clay/cocoa/coreservices/generated/generated.macosx.x86.64.clay [hunk 4]
```diff
@@ -1927,9 +1927,9 @@ alias midiNoClientErr = 4294967046;
 alias midiNoPortErr = 4294967045;
 alias midiTooManyPortsErr = 4294967044;
 alias midiTooManyConsErr = 4294967043;
-alias midiVConnectErr = 4294967042;
-alias midiVConnectMade = 4294967041;
-alias midiVConnectRmvd = 4294967040;
+alias m\ConnectErr = 4294967042;
+alias m\ConnectMade = 4294967041;
+alias m\ConnectRmvd = 4294967040;
```
Comment (jckarter) [inline lib-clay/cocoa/coreservices/generated/generated.macosx.x86.64.clay]: oops

File: compiler/src/lexer.cpp (hunk 1 above)
Comment (jckarter) [inline compiler/src/lexer.cpp]: I don't see how that's possible. `"\\="` and `"\x5c\x3d"` are equivalent.

File: lib-clay/cocoa/coreservices/generated/generated.macosx.x86.64.clay (hunk 4 above)
Comment (ghost) [inline lib-clay/cocoa/coreservices/generated/generated.macosx.x86.64.clay]: Damn it, there's always something when coding late at night . . .

File: compiler/src/lexer.cpp (hunk 1 above)
Comment (ghost) [inline compiler/src/lexer.cpp]: Hmm, yeah thought that was weird, tried it again and it worked fine. 
Must've had a typo or something in there last time.

File: lib-clay/numbers/numbers.clay [hunk 5]
```diff
@@ -153,6 +153,15 @@ inline overload assign(ref dest:A, src:B) {
 }
 
 
+//
+// UpdateAssign \
+//
+
+[A | Integer?(A)]
+inline overload updateAssign(static divide, a:A, b:A) {
```
Comment (jckarter) [inline lib-clay/numbers/numbers.clay]: Did you mean `static quotient` ? Maybe add a test for `\=`.

File: lib-clay/numbers/numbers.clay [hunk 6]
```diff
@@ -274,9 +283,19 @@ inline overload divide(a:A, b:B) =
     divide(toBiggerNumericType(A, B, a),
            toBiggerNumericType(A, B, b));
 
-[A | Numeric?(A)]
-inline overload divide(a:A, b:A) = numericDivide(a, b);
+[A | Float?(A)]
+inline overload divide(a:A, b:A) = floatDivide(a, b);
 
+[A,B | Integer?(A,B)]
+inline overload divide(a:A, b:B) = divide(Float(a), b);
```
Comment (jckarter) [inline lib-clay/numbers/numbers.clay]: I feel like `divide` shouldn't be defined for integers.

File: lib-clay/numbers/numbers.clay [hunk 7]
```diff
@@ -513,7 +535,7 @@ inline wrapSubtract(a:A, b:A) : A = numericSubtract(a, b);
 [A | Integer?(A)]
 inline wrapMultiply(a:A, b:A) : A = numericMultiply(a, b);
 [A | Integer?(A)]
-inline wrapDivide(a:A, b:A) : A = numericDivide(a, b);
+inline wrapDivide(a:A, b:A) : A =  integerDivide(a, b);
```
Comment (jckarter) [inline lib-clay/numbers/numbers.clay]: `wrapDivide` should probably be renamed `wrapQuotient`. The primitive could be renamed to `integerQuotient` too.

**Comment by ghost** [comment]
I think that's it now.

File: compiler/src/codegen.cpp [hunk 8]
```diff
@@ -5355,7 +5359,7 @@ void codegenPrimOp(PrimOpPtr x,
         assert(out->size() == 1);
         CValuePtr out0 = out->values[0];
         assert(out0->type.ptr() == t.ptr());
-        codegenCallValue(staticCValue(operator_doIntegerDivideChecked(), ctx),
+        codegenCallValue(staticCValue(operator_dointegerQuotientChecked(), ctx),
```
Comment (jckarter) [inline compiler/src/codegen.cpp]: Looks like you lowercased the `I` by accident.

**Comment by jckarter** [comment]
Looks good aside from the case change on `doIntegerQuotientChecked`.

**Pull request merged**